karabo_data/geometry2.py

```diff
@@ -631,6 +737,228 @@ def plot_data(self, modules_data):
 {name}/coffset = {coffset}
 """
 
+class LPD_1MGeometry(DetectorGeometryBase):
+    """Detector layout for LPD-1M
+
+    The coordinates used in this class are 3D (x, y, z), and represent multiples
+    of the pixel size.
+
+    You won't normally instantiate this class directly:
+    use one of the constructor class methods to create or load a geometry.
+    """
+    pixel_size = 5e-4  # 5e-4 metres == 0.5 mm
+    frag_ss_pixels = 32
+    frag_fs_pixels = 128
+    expected_data_shape = (16, 256, 256)
+
+    @classmethod
+    def from_quad_positions(cls, quad_pos, *, unit=1e-3, asic_gap=None,
+                            panel_gap=None):
+        """Generate an LPD-1M geometry from quadrant positions.
+
+        This produces an idealised geometry, assuming all modules are perfectly
+        flat, aligned and equally spaced within their quadrant.
+
+        The quadrant positions refer to the corner of each quadrant
+        where module 4, tile 16 is positioned.
+        This is the corner of the last pixel as the data is stored.
+        In the initial detector layout, the corner positions are for the top
+        left corner of the quadrant, looking along the beam.
+
+        The origin of the coordinates is in the centre of the detector.
+        Coordinates increase upwards and to the left (looking along the beam).
+
+        Parameters
+        ----------
+        quad_pos: list of 2-tuples
+          (x, y) coordinates of the last corner (the one by module 4) of each
+          quadrant.
+        unit: float, optional
+          The conversion factor to put the coordinates into metres.
+          The default 1e-3 means the numbers are in millimetres.
+        asic_gap: float, optional
+          The gap between adjacent tiles/ASICs. The default is 4 pixels.
+        panel_gap: float, optional
+          The gap between adjacent modules/panels. The default is 4 pixels.
+        """
+        px_conversion = unit / cls.pixel_size
+        asic_gap_px = 4 if (asic_gap is None) else asic_gap * px_conversion
+        panel_gap_px = 4 if (panel_gap is None) else panel_gap * px_conversion
+
+        tile_size = np.array([cls.frag_fs_pixels, cls.frag_ss_pixels, 0])
+
+        panels_across = [-1, -1, 0, 0]
+        panels_up = [0, -1, -1, 0]
+        modules = []
+        for p in range(16):
+            quad = p // 4
+            quad_corner_x = quad_pos[quad][0] * px_conversion
+            quad_corner_y = quad_pos[quad][1] * px_conversion
+
+            p_in_quad = p % 4
+            # Top beam-left corner of panel
+            panel_corner_x = (quad_corner_x +
+                              (panels_across[p_in_quad] * (
+                                          256 + asic_gap_px + panel_gap_px)))
+            panel_corner_y = (quad_corner_y +
+                              (panels_up[p_in_quad] * (256 + (
+                                          7 * asic_gap_px) + panel_gap_px)))
```

Comment:
Are the 256 coming from the ``expected_data_shape`` tuple? Could you also refer to  them by ``expected_data_shape[1/2]`` rather than writing 256 explicitly?

Reply:
Kind of, but `expected_data_shape` relates to the data we get, rather than the physical layout of the detector itself. I've had a go at clarifying this in a different way.